stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/messages/view/MessageListScrollHelper.kt

```diff
@@ -0,0 +1,155 @@
+package io.getstream.chat.android.ui.messages.view
+
+import androidx.core.view.isVisible
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.recyclerview.widget.RecyclerView
+import com.getstream.sdk.chat.adapter.MessageListItem
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.ui.messages.adapter.BaseMessageItemViewHolder
+import io.getstream.chat.android.ui.messages.adapter.MessageListItemAdapter
+import io.getstream.chat.android.ui.utils.extensions.cast
+import io.getstream.chat.android.ui.utils.extensions.isDeleted
+import io.getstream.chat.android.ui.utils.extensions.safeCast
+import kotlin.math.max
+
+internal class MessageListScrollHelper(
+    private val recyclerView: RecyclerView,
+    private val scrollButtonView: ScrollButtonView,
+    private val scrollToBottomButtonEnabled: Boolean,
+    private val callback: MessageReadListener,
+) {
+    internal var alwaysScrollToBottom = false
+
+    private val layoutManager: LinearLayoutManager by lazy { recyclerView.layoutManager as LinearLayoutManager }
+    private val adapter: MessageListItemAdapter by lazy { recyclerView.adapter as MessageListItemAdapter }
+
+    private var lastSeenMessageInChannel: MessageListItem? = null
+    private var lastSeenMessageInThread: MessageListItem? = null
+    private var isAtBottom = false
+    private var unreadCount = 0
+
+    private var currentList: List<MessageListItem> = emptyList()
+        get() {
+            return adapter.currentList
+        }
+
+    init {
+        scrollButtonView.setOnClickListener {
+            recyclerView.scrollToPosition(currentList.lastIndex)
+        }
+        recyclerView.addOnScrollListener(
+            object : RecyclerView.OnScrollListener() {
+                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
+                    if (!scrollToBottomButtonEnabled || currentList.isEmpty()) return
+
+                    val lastVisibleItemPosition = layoutManager.findLastVisibleItemPosition()
+                    val lastPotentiallyVisibleItemPosition = currentList.indexOfLast { it.isValid() }
+                    val bottomOffset = lastPotentiallyVisibleItemPosition - lastVisibleItemPosition
+                    isAtBottom = bottomOffset == 0
+
+                    max(lastVisibleItemPosition, getLastSeenItemPosition())
+                        .coerceIn(0, currentList.lastIndex)
```

Comment:
Do we need this coerce check? Seems it can't exceed limits

Reply:
There are similar checks in the old `MessageListView`. I guess this was done for a reason to cover some edge case. So I'm inclined to leave it as is for now.